walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,192 @@
+	// Some type of collection to store all of the (unsorted) unavailable meeting times. 
+    Collection<TimeRange> unavailableTimesSet = new HashSet<>();
+
+    /* For each of the requested attendees, loop through all of the events. For each event,
+     * If that particular required attendee is attending it, add that meeting time slot to 
+     * The collection of unavailable meeting times. */
+
+    // Iterate over the HashSet of attendees in request.attendees. 
+    for (String requiredAttendee : request.getAttendees()) {
+      // Loop through all of the events that are going on that day.
+      for (Event currEvent : events) {
+        // Check to see if requiredAttendee is among the attendees at this particular event.
+        if (currEvent.getAttendees().contains(requiredAttendee)) {
+          unavailableTimesSet.add(currEvent.getWhen());
+        }
+      }
+    } 
+
+    /* Sort all the unavailabe meeting times (i.e TimeRanges) in order of start time
+     * (earliest start time first). */
+    List<TimeRange> unfixedSortedUnavailableTimes = new ArrayList<>();
+    unfixedSortedUnavailableTimes.addAll(unavailableTimesSet);
+    Collections.sort(unfixedSortedUnavailableTimes, TimeRange.ORDER_BY_START);
+
+
+    /* See if there's overlap in the unavailable times and fix them in new array. */
+    List<TimeRange> sortedUnavailableTimes = new ArrayList<>();
+
+    if (unfixedSortedUnavailableTimes.size() == 0) {
+      return sortedUnavailableTimes;
+    }
+
+    // if false, last unavailable time in unfixedSortedUnavailableTimes didn't overlap with any before it
+    boolean isLastTimRangeOverlapped = false; 
+    int curr = 0;
+    int next = curr + 1;
+    TimeRange currRange;
+    TimeRange nextRange;
+    while (next != unfixedSortedUnavailableTimes.size()) {
+      currRange = unfixedSortedUnavailableTimes.get(curr);
+      nextRange = unfixedSortedUnavailableTimes.get(next);
+    
+      if (!currRange.overlaps(nextRange)) {
+        sortedUnavailableTimes.add(currRange);
+        curr++;
+        next++;
+        continue;
+      }
+
+      while (currRange.overlaps(nextRange)) {
+        next++;
+        if (next == unfixedSortedUnavailableTimes.size()) {
+          break;
+        } else {
+          nextRange = unfixedSortedUnavailableTimes.get(next);
+        }
+      }
+
+      // If next is now the size of unfixedSortedUnavailableTimes, currRange overlaps 
+      // with all of the following events.
+      if (next == unfixedSortedUnavailableTimes.size()) {
+        isLastTimRangeOverlapped = true;
+        
+        next--;
+        // reset back to the most recent range that overlap.
+        nextRange = unfixedSortedUnavailableTimes.get(next); 
+        
+        // The current range fully contains the last thing it overlaps with.
+        if (currRange.contains(nextRange)) {
+          sortedUnavailableTimes.add(TimeRange.fromStartEnd(currRange.start(), currRange.end(), false));
```

Comment:
If you can create a new function in TimeRange:
TimeRange mergeTimeRange(TimeRange r1, TimeRange r2);
(sorry, in the prev comment, I said that as "operator +" but realized that operator overloading is not supported in Java),
you don't need this if-else but you can do sortedUnavailableTimes.add(mergeTimeRange(currRange, nextRange)).

Reply:
Okay, I'll try implementing that